[Bug]: test_flashinfer_cutlass_mxfp4_fused_moe accuracy mismatch on H20 (sm90) — 89% mismatch vs 20% threshold

### Your current environment

- vllm `0.23.0` (PyPI release) and `vllm-project/vllm@main` (also reproduced on a downstream fork at the same upstream sha)
- flashinfer (whichever version is bundled in the official vllm 0.23.0 wheel)
- transformers: verified on both `4.57.1` and `5.12.1` (latest stable, released 2026-06-15) — same failure
- 1 × NVIDIA H20 (sm_90), CUDA 13.0, torch `2.11.0+cu130`

### 🐛 Describe the bug

`tests/kernels/moe/test_ocp_mx_moe.py::test_flashinfer_cutlass_mxfp4_fused_moe` fails accuracy on Hopper (sm90). 8 parametrize cells all fail with the same shape:

```
test_flashinfer_cutlass_mxfp4_fused_moe[1.0-1.0-None-3072-3072-1-32-1]
test_flashinfer_cutlass_mxfp4_fused_moe[1.0-1.0-None-3072-3072-1-32-4]
test_flashinfer_cutlass_mxfp4_fused_moe[1.0-1.0-None-3072-3072-128-32-1]
test_flashinfer_cutlass_mxfp4_fused_moe[1.0-1.0-None-3072-3072-128-32-4]
test_flashinfer_cutlass_mxfp4_fused_moe[1.702-1.0-7.0-3072-3072-1-32-1]
test_flashinfer_cutlass_mxfp4_fused_moe[1.702-1.0-7.0-3072-3072-1-32-4]
test_flashinfer_cutlass_mxfp4_fused_moe[1.702-1.0-7.0-3072-3072-128-32-1]
test_flashinfer_cutlass_mxfp4_fused_moe[1.702-1.0-7.0-3072-3072-128-32-4]
```

Stack:
```
E   Exception: Mismatch percentage is 0.8997 for rtol 0.3 (threshold: 0.2000)
```

at `check_accuracy(ref, out, atol=0, rtol=0.3, percent=0.8)` after `flashinfer_cutlass_fused_moe(...)` returns.

### Reproducer

```bash
pytest -v tests/kernels/moe/test_ocp_mx_moe.py::test_flashinfer_cutlass_mxfp4_fused_moe
```

The test gates on `HOPPER_MXFP4_BF16_AVAILABLE` (sm90 + flashinfer present), so it is collected on H20.

### Notes / what I checked

- Reproduced on **stock `vllm 0.23.0` PyPI wheel** (with whatever flashinfer ships with it) on an H20. Not a fork-private regression.
- `flashinfer_cutlass_fused_moe` is the flashinfer library kernel; the 89% mismatch likely sits in the flashinfer kernel itself or in how vllm calls it on sm90.
- Searched closed/merged PRs and recent issues; found nothing tracking this. PRs that *look* mxfp4-related (#43557 vllm csrc on B200, #41015 DSv4 indexer kernel) touch different code paths and don't apply here.

### Suggested next steps (no preference, just listing)

1. Pin a known-good flashinfer version that produces correct results on sm90 for this kernel and gate the test on it (or add an autotuning guard, similar in spirit to #37091).
2. If flashinfer doesn't actually support mxfp4 cutlass MoE on sm90, narrow the test gate (currently `HOPPER_MXFP4_BF16_AVAILABLE`) to skip the sm90 path until support lands.
3. Open a corresponding flashinfer-side issue if the kernel is genuinely incorrect on sm90.

Filing this only as a report — happy to dig deeper or put up a PR if a maintainer indicates the preferred direction.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

--- a/tests/kernels/moe/test_ocp_mx_moe.py
+++ b/tests/kernels/moe/test_ocp_mx_moe.py
@@ -8,7 +8,6 @@ from importlib.util import find_spec
 import pytest
 import torch
 from packaging import version
-
 from tests.kernels.moe.utils import check_accuracy
 from vllm._aiter_ops import is_aiter_found
 from vllm.platforms import current_platform
@@ -761,8 +760,8 @@ def test_flashinfer_cutlass_mxfp4_fused_moe(
     # SM90 mixed-input GEMM expects weights/scales in an interleaved layout;
     # without it the FP4->BF16 LUT reads bytes from wrong positions for K>128.
     from flashinfer.fused_moe import (
-        interleave_moe_weights_for_sm90_mixed_gemm,
         interleave_moe_scales_for_sm90_mixed_gemm,
+        interleave_moe_weights_for_sm90_mixed_gemm,
     )
 
     w13_q_swapped = interleave_moe_weights_for_sm90_mixed_gemm(

--- a/tests/kernels/moe/test_ocp_mx_moe.py
+++ b/tests/kernels/moe/test_ocp_mx_moe.py
@@ -8,6 +8,7 @@ from importlib.util import find_spec
 import pytest
 import torch
 from packaging import version
+
 from tests.kernels.moe.utils import check_accuracy
 from vllm._aiter_ops import is_aiter_found
 from vllm.platforms import current_platform

--- a/tests/kernels/moe/test_ocp_mx_moe.py
+++ b/tests/kernels/moe/test_ocp_mx_moe.py
@@ -659,18 +659,6 @@ def test_trtllm_gen_mxfp4_fused_moe(
     check_accuracy(ref_result, tg_result, atol=0, rtol=0.3, percent=0.8)
 
 
-def _interleave_scales_lastdim_by4(scales: torch.Tensor) -> torch.Tensor:
-    """Interleave scales on the last dimension by groups of 4, matching
-    the transformation in mxfp4.py's BF16 (Hopper) path."""
-    s = scales.to(torch.uint8)
-    s_shape = s.shape
-    assert s_shape[-1] % 4 == 0
-    s = s.reshape(*s_shape[:-1], s_shape[-1] // 4, 4)
-    # Move the 4-group dimension before the row dimension
-    permuted = s.permute(0, 2, 1, 3)
-    # Merge the row dim with the 4-group dim
-    return permuted.reshape(s_shape[0], s_shape[-1] // 4, s_shape[1] * 4)
-
 
 @pytest.mark.parametrize("topk", [1, 4])
 @pytest.mark.parametrize("num_experts", [32])
@@ -771,13 +759,24 @@ def test_flashinfer_cutlass_mxfp4_fused_moe(
     w1_w, w3_w = torch.chunk(w13_q, 2, dim=1)
     w13_q_swapped = torch.cat([w3_w, w1_w], dim=1)
 
+    # SM90 mixed-input GEMM expects weights/scales in an interleaved layout;
+    # without it the FP4->BF16 LUT reads bytes from wrong positions for K>128.
+    from flashinfer.fused_moe import (
+        interleave_moe_weights_for_sm90_mixed_gemm,
+        interleave_moe_scales_for_sm90_mixed_gemm,
+    )
+    w13_q_swapped = interleave_moe_weights_for_sm90_mixed_gemm(
+        w13_q_swapped, quant_type="fp4"
+    )
+    w2_q = interleave_moe_weights_for_sm90_mixed_gemm(w2_q, quant_type="fp4")
+
     b1, b3 = torch.chunk(bias13.to(torch.float32), 2, dim=-1)
     w13_b = torch.cat([b3, b1], dim=-1).to(torch.bfloat16)
 
     w1_s, w3_s = torch.chunk(w13_scale, 2, dim=1)
     w13_s = torch.cat([w3_s, w1_s], dim=1)
-    w13_s_inter = _interleave_scales_lastdim_by4(w13_s)
-    w2_s_inter = _interleave_scales_lastdim_by4(w2_scale)
+    w13_s_inter = interleave_moe_scales_for_sm90_mixed_gemm(w13_s)
+    w2_s_inter = interleave_moe_scales_for_sm90_mixed_gemm(w2_scale)
 
     routing_weights = torch.nn.functional.softmax(
         router_logits, dim=1, dtype=torch.float32

--- a/tests/kernels/moe/test_ocp_mx_moe.py
+++ b/tests/kernels/moe/test_ocp_mx_moe.py
@@ -659,7 +659,6 @@ def test_trtllm_gen_mxfp4_fused_moe(
     check_accuracy(ref_result, tg_result, atol=0, rtol=0.3, percent=0.8)
 
 
-
 @pytest.mark.parametrize("topk", [1, 4])
 @pytest.mark.parametrize("num_experts", [32])
 @pytest.mark.parametrize("num_tokens", [1, 128])
@@ -765,6 +764,7 @@ def test_flashinfer_cutlass_mxfp4_fused_moe(
         interleave_moe_weights_for_sm90_mixed_gemm,
         interleave_moe_scales_for_sm90_mixed_gemm,
     )
+
     w13_q_swapped = interleave_moe_weights_for_sm90_mixed_gemm(
         w13_q_swapped, quant_type="fp4"
     )

